### PR TITLE
ref: Allow auto-assignment after manual assignment if called from API

### LINKED ISSUE #52318: Create an API which will force run auto-assignment
We built a feature which prevents auto-assignment of an issue which was manually assigned.  This is to prevent us from incorrectly auto-assigning an issue when it has been determined that there is a better assignee.  This however can lead to an issue where if issues were accidentally assigned in bulk, that we don't have a way to allow auto-assignment to fix it.  

AC:
- Given a list of group IDs, force auto-assignment to run based on auto-assignment configuration for that project.  
- Limit this API to 100 group IDs per request
- Implement a rate limit of one request per minute per org

### PR DESCRIPTION
**Pull request by jangjodi**
Allow auto-assignment after manual assignment if called from force auto-assignment API (see https://github.com/getsentry/getsentry/pull/11401)

closes https://github.com/getsentry/sentry/issues/52318

### CODE REVIEW

File: src/sentry/api/endpoints/organization_force_auto_assignment.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+@region_silo_endpoint
+class OrganizationForceAutoAssignmentEndpoint(OrganizationEndpoint):
+
+    rate_limits = {"PUT": {RateLimitCategory.ORGANIZATION: RateLimit(1, 60)}}  # 1 rpm
+
+    def put(self, request: Request, organization: Organization) -> Response:
+        """
+        Endpoint for forcing autoassignment to run for specified group ids.
+        This is for if a user incorrectly manually assigns a group and wants autoassignment to run.
```
Comment (sentaur-athena) [inline src/sentry/api/endpoints/organization_force_auto_assignment.py:22]: This seems like a parameter addition to an existing api instead of a new one. Can you point me to the assign api?

File: src/sentry/api/endpoints/organization_force_auto_assignment.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+from sentry.api.bases.organization import OrganizationEndpoint
+from sentry.api.serializers import serialize
+from sentry.models import Organization
+from sentry.models.group import Group
+from sentry.models.projectownership import ProjectOwnership
+from sentry.types.ratelimit import RateLimit, RateLimitCategory
+
+
+@region_silo_endpoint
+class OrganizationForceAutoAssignmentEndpoint(OrganizationEndpoint):
```
Comment (sentaur-athena) [inline src/sentry/api/endpoints/organization_force_auto_assignment.py:15]: This seems like an endpoint that could be called directly by user. Please make sure to include documentation and publish it. Here's an example: https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/organization_member/details.py#L90

File: src/sentry/api/endpoints/organization_force_auto_assignment.py (hunk 1 above)
Comment (jangjodi) [reply src/sentry/api/endpoints/organization_force_auto_assignment.py:22]: There is a manual assign API [here](https://github.com/getsentry/sentry/blob/1cb4428d34d556792e166d0119eca7f41cbe42e7/src/sentry/api/endpoints/group_details.py#L277), but this API is different as we want to call auto-assignment. Auto-assignment is usually called in a task [here](https://github.com/getsentry/sentry/blob/1cb4428d34d556792e166d0119eca7f41cbe42e7/src/sentry/tasks/post_process.py#L981) and not in an API

File: src/sentry/api/endpoints/organization_force_auto_assignment.py (hunk 2 above)
Comment (jangjodi) [reply src/sentry/api/endpoints/organization_force_auto_assignment.py:15]: We want the endpoint to be directly called by a user, but do not wish to make it public as this can be a labour-intensive call (see [here](https://sentry.slack.com/archives/C04KZQBNQ2U/p1688594072303679?thread_ts=1687793787.448889&cid=C04KZQBNQ2U)). This API will only be called by users who reach out to support and ask for the functionality.

(cc: @NisanthanNanthakumar @brianthi)

Comment (sentaur-athena) [reply src/sentry/api/endpoints/organization_force_auto_assignment.py:15]: Thanks for the context. Going towards the goal of API first company we prefer all public APIs to be publicly documented so let's discuss other solutions. Also if it's not public how do they know about it to ask from support?!
1. From the conversation seems like excessive rate limiting is a good idea to address this issue. Does it not?
2. What is the limit? How many orgs can have access to this and why does that limit exists? If you provide more context on the technical challenge maybe we can find a better solution. Hiding it definitely won't work, what if too many orgs find out about it through support to just code?

File: src/sentry/api/endpoints/organization_force_auto_assignment.py (hunk 1 above)
Comment (sentaur-athena) [reply src/sentry/api/endpoints/organization_force_auto_assignment.py:22]: > Auto-assignment is usually called in a task [here](https://github.com/getsentry/sentry/blob/1cb4428d34d556792e166d0119eca7f41cbe42e7/src/sentry/tasks/post_process.py#L981) and not in an API

Thanks for the pointers. You're calling it in an api now so it seems like a good fit for the API. For example can assignedTo param get user:AUTO_ASSIGN or even add a new enum that assignmentMethod

File: src/sentry/api/endpoints/organization_force_auto_assignment.py (hunk 2 above)
Comment (jangjodi) [reply src/sentry/api/endpoints/organization_force_auto_assignment.py:15]: I think this would work. @sentaur-athena what do you think?

For more context, a user would only want to use this API if they manually bulk assigned groups after auto-assignment, then later wanted auto-assignment to run. In general, users can manually assign groups if they made a mistake in assignment. This API was made in response to a customer who had the first use case and reached out to support asking if this was possible. 

Excessive rate limiting could address this issue, but I think it's a rare enough use case that we don't need this labor intensive API to be public.

File: src/sentry/models/projectownership.py [hunk 3]
```diff
@@ -251,6 +251,11 @@ def handle_auto_assignment(cls, project_id, event):
             User,
         )
 
+        force_autoassign = True
```
Comment (armenzg) [inline src/sentry/models/projectownership.py:254]: Could you please add a comment to this block? It's hard to follow as to why an event would cause auto assignment or not.

**Review by armenzg — CHANGES_REQUESTED** [review]
It seems that one of the two cases in here is not being tested (assuming my codecov add-on is loading the right data for this PR):
image

**Review by armenzg — APPROVED** [review]
Thank you @jangjodi
